src/Bicep.LangServer/Refactor/TypeStringifier.cs

```diff
@@ -0,0 +1,294 @@
+            if (nonNullMembers.Select(m => WidenLiteralType(m).Name).Distinct().Count() == 1) // using "Name" comparison because we don't care about ValidationFlags
+            {
+                var widenedType = WidenLiteralType(nonNullMembers[0]);
+                if (TypeHelper.IsNullable(unionType))
+                {
+                    // If it had "|null" before, add it back
+                    widenedType = TypeHelper.MakeNullable(widenedType);
+                }
+                return widenedType;
+            }
+            else
+            {
+                return type;
+            }
+        }
+
+        // ... otherwise (medium and loose) just widen literals to simple types
+        return type switch
+        {
+            StringLiteralType => LanguageConstants.String,
+            IntegerLiteralType => LanguageConstants.Int,
+            BooleanLiteralType => LanguageConstants.Bool,
+            _ => type,
+        };
+    }
+
+    private static string Arrayify(TypeSymbol type, Strictness strictness, TypeSymbol[] visitedTypes)
+    {
+        string stringifiedType = StringifyCore(type, null, strictness, visitedTypes);
+        bool needsParentheses = NeedsParentheses(type, strictness);
+
+        return needsParentheses ? $"({stringifiedType})[]" : $"{stringifiedType}[]";
+    }
+
+    private static string Nullableify(TypeSymbol type, Strictness strictness, TypeSymbol[] visitedTypes)
+    {
+        string stringifiedType = StringifyCore(type, null, strictness, visitedTypes);
+        bool needsParentheses = NeedsParentheses(type, strictness);
+
+        return needsParentheses ? $"({stringifiedType})?" : $"{stringifiedType}?";
+    }
+
+    private static bool NeedsParentheses(TypeSymbol type, Strictness strictness)
+    {
+        // If the type is '1|2', with loose/medium, we need to check whether 'int' needs parentheses, not '1|2'
+        // Therefore, widen first
+        var widenedType = WidenType(type, strictness);
+        bool needsParentheses = widenedType switch
+        {
+            UnionType { Members.Length: > 1 } => true, // also works for nullable types
+            _ => false
+        };
+        return needsParentheses;
+    }
+
+    private static bool IsLiteralType(TypeSymbol type) => WidenLiteralType(type) != type;
+
+    private static TypeSymbol WidenLiteralType(TypeSymbol type) => type switch
+    {
+        // Perhaps we should have a LiteralType class
+        StringLiteralType => LanguageConstants.String,
+        BooleanLiteralType => LanguageConstants.Bool,
+        IntegerLiteralType => LanguageConstants.Int,
+        _ => type
+    };
+
+    private static IEnumerable<TypeSymbol> NonNullUnionMembers(UnionType unionType) =>
+        unionType.Members.Where(m => m.Type is not NullType).Select(tr => tr.Type);
+
+    // True if "{}" (which allows no additional properties) instead of "object"
+    private static bool IsObjectLiteral(ObjectType objectType)
```

Comment:
This is a minor nitpick, but I’d suggest naming it `IsEmpty` or `IsEmptyObjectLiteral`, as an object literal can still contain properties

Reply:
makes sense